pkg/limit/fixed_window_ddb.go

```diff
@@ -0,0 +1,100 @@
+package limit
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/justtrackio/gosoline/pkg/cfg"
+	"github.com/justtrackio/gosoline/pkg/clock"
+	"github.com/justtrackio/gosoline/pkg/ddb"
+	"github.com/justtrackio/gosoline/pkg/log"
+)
+
+const (
+	attrVal = "val"
+	attrTtl = "ttl"
+)
+
+type window struct {
+	Key string `json:"key" ddb:"key=hash"`
+	Val int    `json:"val"`
+	Ttl int64  `json:"ttl" ddb:"ttl=enabled"`
+}
+type fixedWindowDdb struct {
+	clock  clock.Clock
+	repo   ddb.Repository
+	tRepo  ddb.TransactionRepository
+	window time.Duration
+	name   string
+	logger log.Logger
+}
+
+func NewFixedWindowDdb(ctx context.Context, config cfg.Config, logger log.Logger, settings *ddb.Settings, c FixedWindowConfig) (LimiterWithMiddleware, error) {
+	settings.Main.Model = &window{}
+
+	repo, err := ddb.NewRepository(ctx, config, logger, settings)
+	if err != nil {
+		return nil, err
+	}
+
+	transactionRepo, err := ddb.NewTransactionRepository(ctx, config, logger, "default")
+	if err != nil {
+		return nil, err
+	}
+
+	backend := &fixedWindowDdb{
+		logger: logger.WithChannel("rate_limiter_incrementer_ddb"),
+		clock:  clock.NewRealClock(),
+		repo:   repo,
+		tRepo:  transactionRepo,
+		window: c.Window,
+		name:   c.Name,
+	}
+
+	builder, err := newInvocationBuilder(c.Name)
+	if err != nil {
+		return nil, err
+	}
+
+	return NewFixedWindowLimiter(backend, clock.NewRealClock(), c, builder), nil
+}
+
+func (f fixedWindowDdb) Increment(ctx context.Context, prefix string) (incr *int, ttl *time.Duration, err error) {
+	key := f.keyBuilder(prefix)
+
+	item := &window{Key: key}
+
+	now := f.clock.Now
+
+	reset := f.repo.UpdateItemBuilder().
+		WithCondition(ddb.Or(
+			ddb.AttributeNotExists(attrTtl),
+			ddb.Lt(attrTtl, now().Unix()))).
+		Set(attrVal, 1).
+		Set(attrTtl, now().Add(f.window).Unix()).
+		ReturnAllNew()
+
+	resp, err := f.repo.UpdateItem(ctx, reset, item)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	if resp.ConditionalCheckFailed {
+		increment := f.repo.UpdateItemBuilder().
+			Add(attrVal, 1).
```

Comment:
Do you really need this second request?

DDB has the SetIfNotExists capability, and you can call Add on a non-existent property and it will use the the default value (for numbers: 0)

Reply:
So, your previous Update could look like this:

```
 f.repo.UpdateItemBuilder().
		WithCondition(ddb.Or(
			ddb.AttributeNotExists(attrTtl),
			ddb.Lt(attrTtl, now().Unix()))).
		Add(attrVal, 1).
		SetIfNotExists(attrTtl, now().Add(f.window).Unix()).
		ReturnAllNew()
```

You need to adapt the conditional check logic then; in the new case it would mean that the element already has expired and we're in a new window, so directly returning with the permission to execute is fine?